examples/nlp/gpt/conf/gpt_spin.yaml

```diff
@@ -61,6 +61,7 @@ model:
 
   spin:
     log_prob_forward_micro_batch_size: 1
+    generation_batch_size: 16
```

Comment:
So as I understand it (and as @gshennvm explained it to me), when you call `model.generate` it is the shape of the input tensor being passed which determines the generation MBS. So if you pass a tensor of shape (2, 4096) then your generation MBS is that first dimension of 2. There is no explicit `model.generate` mbs parameter, it's inferred from the input tensor shape.

Prior to this re-write, the tensor which was passed to `model.generate` was a single batch pulled from the train dataloader, and the shape of that was GBS // DP. This was inefficient and the user had no control over it.

The new approach in this PR allows for several draws/pulls from the train dataloader on each DP rank in order to build up the ultimate local batch which is sent to generation, such that it has shape (X, 4096), where X is the hyperparam added by this PR.

Reply:
> it has shape (X, 4096), where X is the hyperparam added by this PR.

Doesn't it have shape X * GBS // DP?

Reply:
No, it's X, what it does is store (GBS // DP) batches pulled from train dataloader into a buffer until it has exactly X size, then it sends that to model.generate

Reply:
Ooops, I was wrong, it was shape X * GBS // DP, I have now corrected it so it is checking correctly that the shape is X

Reply:
_Discussed offline, we now both agree that the current implementation yields to generating batches of length X * GBS // DP._
 
Here's what I would suggest in terms of implementation:
* Rename `generation_batch_size` into `rollout_micro_batch_size` (for consistency with PPO naming)
* This setting should either divide `gbs // dp` or be a multiple of it (otherwise raise an exception). The former is useful for large `gps // dp` to avoid OOM during generation, while the latter is useful for small `gbs // dp` to maximize efficiency.
* In the `while not done:` loop of `augment_dataloader()`, do only one iteration in the former case, and otherwise a number of iterations equal to `rollout_micro_batch_size // (gbs // dp)`
* Modify `get_generations()` to make multiple calls to `generate()` in the former case.

NB: the newest commit b763797bca8bdeef70f20cc7f3e5c55c71780829 that was pushed while I was writing this handles the latter case only, but not the former (which I think we should also support).

Reply:
Resolving as per https://github.com/NVIDIA/NeMo-Aligner/pull/133#pullrequestreview-2038004886